design-document/design-document.md

```diff
@@ -0,0 +1,131 @@
+ - sort and filter on all available attributes (`name`, `size`, and `type`)
+ - bookmark URLs for quick navigation
+
+On the backend, this entails a simple `server` which:
+ 
+ - serves web assets
+ - manages user sessions
+ - returns contents of a directory on a given path
+
+Finally, a few constraints and allowed/encouraged shortucts:
+
+ - containerize the app (Docker)
+ - use native browser API for networking
+ - use native/standard libraries
+ - store users & sessions in memory, no need for a db
+ - feel free to hardcode usernames, hashes, or directory paths
+ - make it secure (strong TLS, no web security vulnerabilities)
+
+# Wireframes
+
+_Note: Honestly, these are more colourful and specific than wireframes need to be — or maybe even should be — but it was too fast & easy with [Whimsical](https://whimsical.com) to do it any other way._
+
+<img src="assets/signin.png" style="max-width: 600px" alt="Sign In" />
+<img src="assets/browse.png" style="max-width: 600px;" alt="Browse a Directory" /> 
+
+# Implementation Details
+
+The entire codebase shall use typescript, reasonable linting, and be dockerized.
+
+In a real app, error reporting would be done through an integration (like Sentry), but for the purposes of the challenge `console.error` will suffice. 
+
+There will be a few tests for the key React components and for the key API endpoints, but coverage shouldn't become an obsession.
+
+## Frontend
+
+Let's use React (`create-react-app` for the simplicity of its setup), React Router, and Tailwind on the frontend. Core considerations:
+
+ - **Authentication:** All protected routes will be wrapped in a simple component `<Protected />` which checks whether the user is authenticated. When they're not, the user is taken to the `/sign-in` page, and the original URL is stored in the `state` property on the `location` object of our router. That way, we can redirect the user back to where they intended to go once they've signed in.
+ - **URL as the source of truth:** We will store user's current working directory inside the file tree as URL params, the sorting and filtering criteria as query params (`/browse/my-directory/my-child-directory?sort=name-up&search=holida`).
+ - **Error handling**: Let's pay attention to loading states, errors (at the very least 400, 401, 403, 404, 500) and edge cases such as invalid filtering options, etc.
+ - **Security:** Input and output shall be sanitised (URL params, query params, and the search input field).
+
+## Backend
+
+The server will be a plain `express` app. For the sake of simplicity, let's skip a database and store everything in memory instead.
+
+Since searching and filtering is required on the client only, a handful of API endpoints will suffice:
+
+ - `GET /contents?path={path}` to obtain 1-level deep contents of a directory, returning `{ contents: Item }` where `interface Item { name: string, sizeKb: number | undefined, type: "file" | "dir", items: undefined | Item[] }`
```

Comment:
Just curious why not use an URI path instead of query parameters?

Reply:
Ehm, I didn't go anywhere near deep on this, a query parameter just seemed much simpler than going into wildcards or having to directly get the values from `req.url`: https://github.com/expressjs/express/issues/699.

I looked again now, and it turns out it's much simpler: https://stackoverflow.com/questions/45500439/variable-number-of-route-params-in-express. You can specify the handler as `contents/*` and then get the params as a slightly misshapen array-like object `{ params: { '0': 'one/two/three' } }`.

Do you think there's a benefit to implementing it that way?